### ISSUE TITLE
Feature Request: Integrate with Cloud SQL Proxy

### ISSUE THREAD

**Issue by alexburlton-sonocent** [terraform-providers/terraform-provider-postgresql#110]
### Description

It would be great to be able to configure a Postgres instance hosted in GCloud, where access can only be granted using the [Cloud SQL Proxy](https://cloud.google.com/sql/docs/postgres/sql-proxy). Something like this was raised with the google provider already, but was closed (rightly) as it would need to be implemented in this provider instead. In particular, see this comment: https://github.com/terraform-providers/terraform-provider-google/issues/2094#issuecomment-425688564

Currently, the only solutions available to us that I can see are:

 - Configure postgres by hand after terraform has run (not declarative), OR
 - Make the postgres provider point at localhost, and have a requirement that the cloud SQL proxy is running locally whenever you want to run a `terraform apply` (not ideal)

### Terraform Configuration Files
For example, I want to be able to reference my cloud SQL instance in the provider (probably passing some additional details) and have it automatically know to connect using the cloud SQL proxy. Something like:

```hcl
provider postgresql {
  host = google_sql_database_instance.my_instance.???
  port = 5432
  database = google_sql_database.my_instance.name
  username = google_sql_user.my_administrator.name
  password = google_sql_user.my_administrator.password
}
```

### References
- https://github.com/terraform-providers/terraform-provider-google/issues/2094

**Comment by sterlingdeng** [comment]
Agree. I've ran into the same issue but will most likely solve this with #1 which you listed above (doing it manually).

**Comment by mltsy** [comment]
This would be super useful.  I thought about trying to set this up with a `local_exec` provisioner, but ... it seemed pretty messy to get it setup and torn down correctly for a config where I need to connect to and configure several separate postgres servers.

**Comment by tun0** [comment]
>  I have now got it up and running in here [benfdking#1](https://github.com/benfdking/terraform-provider-postgresql/pull/1) with a very simple example. Requires terraform 0.12, couldn't get the setup with the plugins to work with terraform 0.13. It definetely needs some tidying but works pretty much as expected.

Awesome! Do you plan to (continue to) work on 0.13 support?

**Comment by benfdking** [comment]
@tun0 I don't think it's necessarily the fact that it doesn't work in terraform 0.13. In fact there is nothing that to me would indicate it shouldn't: I just couldn't get it setup correctly with the new provider definition, whereas in 0.12 I just placed it in the root folder of where it runs. 

We are using it in production and have placed it [here](https://github.com/tumelohq/terraform-provider-postgresql/) for the time being.  Our aim is to keep it up to date with the added functionality of being able to proxy seamlessly to Google & Amazon Postgres instances. It does still need a bit of love to tidy it up but definetely functional from our limited experience with it.

**Comment by tun0** [comment]
After reading up a bit on the subject, it seems that as of 0.13 using custom provider got quite a bit more complex indeed. Being rather new to Terraform in general, I doubt I could make it work either :wink:

**Comment by benfdking** [comment]
For use with terraform 0.13 and just generally easier use https://registry.terraform.io/providers/tumelohq/postgresql/latest 

Note that the documentation isn't there yet, the best I can do is point to the example.

**Comment by tun0** [comment]
Awesomeness! Gonna take it for a spin when I find some spare minutes. The provided example should be enough to get started.

**Comment by ademariag** [comment]
@benfdking is the intention to have a PR and merge into this provider?

**Comment by benfdking** [comment]
Hey @ademariag 

With the ever so slight bias of being part of the @tumelohq team that created it, I do think this solution is a nice step forward and it has given us oodles of flexibility which could benefit others, among them:
- it feels cloud agnostic
- we can test the setup locally against databases
- none of have a background really in database administration and so this feels more natural with what we know and love with terraform
-  it's one less tool we have to learn or build a script for

From my point of our point of view, I think we would be very proud if this concept would be merged into the provider. It would force us slighlty to polish it up or maybe even start fresh, but we think it would for the better and would make it more accessible to others. 

That said I don't really think that's our call to make and would defer to the maintainers: I am sure they have many more difficult considerations that go way past what I have mentioned. If they feel like this would be valuable for them though, I and I am sure some others on the team would gladly spend some time to help make that a reality.

**Comment by ademariag** [comment]
@benfdking understood thank you for your response.

I think it's still worth creating a pull request and give an opportunity to the maintainers to weight in and decide whether they are ok with it or not 😊
